src/coreclr/vm/comdelegate.cpp

```diff
@@ -1717,6 +1717,11 @@ extern "C" void QCALLTYPE Delegate_Construct(QCall::ObjectHandleOnStack _this, Q
     if (Nullable::IsNullableType(pMeth->GetMethodTable()))
         COMPlusThrow(kNotSupportedException);
 
+    // Do not allow static methods with [UnmanagedCallersOnlyAttribute] to be a delegate target.
```

Comment:
Most delegate constructor go via optimized path, so this is incomplete.

Our strategy has been to delay this check until the UnmanagedCallersOnly method is actually called and produce a fail-fast with a good error message in that case: https://github.com/dotnet/runtime/blob/a83d76a82cd0ef5fa0a8a415fe1d80681122936e/src/coreclr/vm/jithelpers.cpp#L997

Reply:
Where is this check on the optimized path?

Reply:
It's in the optimized delegate creation path called by the jit. I could replicate that if needed in the interpreter compiler. Search for the comment. I copied it verbatim.

Reply:
I see it now. I think it is an oversight that we check for it on the fast path. It is inconsistent with the lazy check that we do everywhere else. For example, you can create delegate pointing to UMCO via reflection just fine (it will fail once you try to call it):
```
public class Program
{
    static void Main()
    {
        typeof(Program).GetMethod("Test").CreateDelegate<Action>();
    }

    [UnmanagedCallersOnly]
    public static void Test() { }
}
```